### PR TITLE
Enable ArrayBoundsCheckElimination on OSSA

### PR DESCRIPTION
**Pull request by meg-gupta**
Minor refactoring in ArrayBoundsCheckElimination
Support hoisting/copying owned/guaranteed self in ArraySemantics
Support OSSA instructions load [copy], copy_value, destroy_value in ArrayBoundsCheckElimination

### CODE REVIEW

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

**Comment by meg-gupta** [comment]
@swift-ci test source compatibility

**Comment by meg-gupta** [comment]
@swift-ci test macOS Platform

File: lib/SILOptimizer/Analysis/ArraySemantic.cpp [hunk 1]
```diff
@@ -376,7 +386,27 @@ static SILValue copyArrayLoad(SILValue ArrayStructValue,
     InsertPt = Inst;
   }
 
-  return cast<LoadInst>(LI->clone(InsertBefore));
+  if (ArrayStructValue->getFunction()->hasOwnership()) {
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/ArraySemantic.cpp:389]: This block of code would be a bit easier to follow like this
```
  if (!ArrayStructValue->getFunction()->hasOwnership()) {
    return cast<LoadInst>(ArrayStructValue)->clone(InsertBefore);
  }
  // followed by all the nested code blocks for the OSSA logic
```

File: lib/SILOptimizer/Analysis/ArraySemantic.cpp [hunk 2]
```diff
@@ -376,7 +386,27 @@ static SILValue copyArrayLoad(SILValue ArrayStructValue,
...
+      return SILBuilderWithScope(InsertPt).createLoad(
+          InsertPt->getLoc(), Addr, LoadOwnershipQualifier::Copy);
+    }
+    if (auto *Load = dyn_cast<LoadInst>(ArrayStructValue)) {
+      Load->moveBefore(InsertBefore);
+      return ArrayStructValue;
+    }
+    auto *Copy = cast<CopyValueInst>(ArrayStructValue);
+    auto Addr = cast<LoadInst>(Copy->getOperand())->getOperand();
+    Copy->replaceAllUsesWithUndef();
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/ArraySemantic.cpp:405]: I can't understand how `replaceAllUsesWithUndef` works here. What about other uses of the array value? The call may take a `guaranteed` self, or may be in a different block from the copy, no?

File: lib/SILOptimizer/Analysis/ArraySemantic.cpp [hunk 3]
```diff
@@ -513,9 +551,15 @@ ApplyInst *swift::ArraySemanticsCall::hoistOrCopy(SILInstruction *InsertBefore,
...
-    SILBuilderWithScope Builder(SemanticsCall);
-    Builder.createReleaseValue(SemanticsCall->getLoc(), getSelf(),
-                               Builder.getDefaultAtomicity());
+    if (!SemanticsCall->getFunction()->hasOwnership()) {
+      SILBuilderWithScope Builder(SemanticsCall);
+      Builder.emitDestroyValueOperation(SemanticsCall->getLoc(), getSelf());
+    } else {
+      auto Self = getSelf();
+      assert(isa<CopyValueInst>(Self) || isa<LoadInst>(Self));
+      Self->replaceAllUsesWithUndef();
```
Comment (atrick) [inline lib/SILOptimizer/Analysis/ArraySemantic.cpp:560]: Again I'm not sure how `replaceAllUsesWithUndef` works in general. Is it assuming the call takes an owned self and is in the same blocks. Even then, what about other non-consuming uses of the array that aren't being rewritten?

**Review by atrick — COMMENTED** [review]
I'm confused about `replaceAllUsesWithUndef` but otherwise it looks good.

File: lib/SILOptimizer/Analysis/ArraySemantic.cpp (hunk 3 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Analysis/ArraySemantic.cpp]: Thanks for pointing it out. I fixed this and added a test.

File: lib/SILOptimizer/Analysis/ArraySemantic.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Analysis/ArraySemantic.cpp]: Fixed it.

**Comment by meg-gupta** [comment]
@swift-ci test

**Review by atrick — APPROVED** [review]
LGTM

**Pull request merged**